rstudio/rstudio
# Electron will-navigate event handling

### Intent
Address #11252 to check whether navigation to the url is accepted

### Approach
Implement similar logic to `WebPage::acceptNavigationRequest`. URL is checked for:

* malformed URL logs an error, no navigation
* `about:blank` and `chrome://gpu` are allowed
* protocols other than `http`, `https`, `mailto`, and `data` are rejected
* no base url and local is allowed
* matching base url is allowed
* viewer, presentation, tutorial, or Shiny dialog is allowed
* allow if external navigation is allowed or is a safe host
* last resort, opens in a browser

### Automated Tests
Unit tests included cover the util methods that have been added and the `allowNavigation` method that is called for the `will-navigate` event.

### QA Notes
Check the Tutorial, Viewer, and Presentation panes and run a Shiny app in a window and in the Viewer pane.

Opening a new project should have the IDE re-load with the project.

### Checklist
